drivers/sensor/sun8i-thermal.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <debug.h>
+#include <error.h>
+#include <limits.h>
+#include <mmio.h>
+#include <sensor.h>
+#include <sensor/sun8i-thermal.h>
+
+#define SENSOR_COUNT                3
+
+#define THS_0_DATA_REGISTER         0x80
+#define THS_1_DATA_REGISTER         0x84
+#define THS_2_DATA_REGISTER         0x88
+
+#define THS_CTL_REGISTER0           0x00
+#define THS_CTL_REGISTER1           0x04
+#define THS_CTL_REGISTER2           0x40
+
+#define THS_FILTER_CONTROL_REGISTER 0x70
+
+enum {
+	THS0,
+	THS1,
+	THS2,
+};
+
+static struct sensor_info sun8i_thermal_sensor_info[SENSOR_COUNT] = {
+	[THS0] = {
+		.name    = "ths0",
+		.reading = UINT32_MAX, /* Symbolizes no reading yet. */
+		.formula = {
+			.min_value  = 680,
+			.multiplier = -8.56,
+		},
+		.address = 0x80,
+	},
+	[THS1] = {
+		.name    = "ths1",
+		.reading = UINT32_MAX, /* Symbolizes no reading yet. */
+		.formula = {
+			.min_value  = 680,
+			.multiplier = -8.56,
+		},
+		.address = 0x84,
+	},
+	[THS2] = {
+		.name    = "ths2",
+		.reading = UINT32_MAX, /* Symbolizes no reading yet. */
+		.formula = {
+			.min_value  = 680,
+			.multiplier = -8.56,
+		},
+		.address = 0x88,
+	},
+};
+
+static struct sensor_info *
+sun8i_thermal_get_info(struct device *dev __unused, uint8_t id)
+{
+	assert(id < SENSOR_COUNT);
+
+	return &sun8i_thermal_sensor_info[id];
+}
+
+static uint32_t
+sun8i_thermal_read_raw(struct device *dev, uint8_t id, uint32_t *raw __unused)
+{
+	uint32_t value;
+
+	value =
+		mmio_read32(dev->regs +
+		            sun8i_thermal_sensor_info[id].address) &
+		BITMASK(0, 11);
+
+	return value;
+}
+
+static int
+sun8i_thermal_probe(struct device *dev)
+{
+	uint32_t reg;
+
+	/* Set up two clocks for the thermal sensor */
+	dm_setup_clocks(dev, 2);
+
+	/* Start calibration. */
+	mmio_write32(dev->regs + THS_CTL_REGISTER1, BIT(17));
+	while (mmio_read32(dev->regs + THS_CTL_REGISTER1) & BIT(17)) {
+		/* Wait for for calibration to clear */
+	}
+
+	/* Set aquire times. */
+	mmio_write32(dev->regs + THS_CTL_REGISTER0, 0x190);
+	mmio_write32(dev->regs + THS_CTL_REGISTER2, 0x190 << 16);
+
+	/* Enable filter, average over 8 values. */
```

Comment:
If you copy code from somewhere else, make sure you respect the license! (Hint: you need another copyright line at the top of the file)

Reply:
done